subsys/net/lib/dns/dns_pack.c

```diff
@@ -404,3 +406,168 @@ int dns_copy_qname(u8_t *buf, u16_t *len, u16_t size,
 
 	return rc;
 }
+
+int dns_unpack_query_header(struct dns_msg_t *msg, u16_t *src_id)
+{
+	u8_t *dns_header;
+	u16_t size;
+	int qdcount;
+	int rc;
+
+	dns_header = msg->msg;
+	size = msg->msg_size;
+
+	if (size < DNS_MSG_HEADER_SIZE) {
+		return -ENOMEM;
+	}
+
+	if (dns_header_qr(dns_header) != DNS_QUERY) {
+		return -EINVAL;
+	}
+
+	if (dns_header_opcode(dns_header) != DNS_QUERY) {
+		return -EINVAL;
+	}
+
+	if (dns_header_z(dns_header) != 0) {
+		return -EINVAL;
+	}
+
+	rc = dns_header_rcode(dns_header);
+	if (rc != 0) {
+		return -EINVAL;
+	}
+
+	qdcount = dns_unpack_header_qdcount(dns_header);
+	if (qdcount < 1) {
+		return -EINVAL;
+	}
+
+	if (src_id) {
+		*src_id = dns_unpack_header_id(dns_header);
+	}
+
+	msg->query_offset = DNS_MSG_HEADER_SIZE;
+
+	return qdcount;
+}
+
+/* Returns the length of the unpacked name */
+static int dns_unpack_name(const u8_t *msg, int maxlen, const u8_t *src,
+			   struct net_buf *buf, const u8_t **eol)
+{
+	const u8_t *curr_src = src;
+	const u8_t *end_of_label = NULL;
+	int dest_size = net_buf_tailroom(buf);
+	int loop_check = 0, len = -1;
+	int label_len;
+	int val;
+
+	if (curr_src < msg || curr_src >= (msg + maxlen)) {
+		return -EMSGSIZE;
+	}
+
+	while ((val = *curr_src++)) {
+		if (val & 0xc0) {
```

Comment:
would you have a macro name for 0xc0 instead? something that self-describe if possible the value.